Option to Pull Parent, Child, and Extension Cases to Case Search

Jtang-1 opened the pull request:
## Product Description
UI Addition: When feature flag is turned on and case search is activated - the toggle for "Pull All Related Cases" is available 
under the module's Case List tab -> Search and Claim Options section.

![image](https://user-images.githubusercontent.com/88759246/224782310-992b5d89-5a83-4d24-89db-69d4471074b2.png)

As of current, with [Case Detail Nodeset](https://confluence.dimagi.com/display/saas/Case+Detail+Nodesets), when a child is defined as a Data Tab, child case types are pulled into case search. When this new feature is turned on, the logic tied with Case Detail Nodeset (as described previously) is overwritten. The new logic will pull in all parent, child, and extension cases related to the base case search "search results".

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[Jira Ticket](https://dimagi-dev.atlassian.net/browse/USH-2575)
[Tech Spec](https://docs.google.com/document/d/1wr_PgRQ6iq-vHvC7TmvrjHJySgeAYTYjyOxlabagyvE/edit#) See "Alternate Proposal: Sync Algorithm"

## Feature Flag
https://www.commcarehq.org/hq/flags/edit/case_claim_autolaunch/

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Local testing that case search behavior unchanged when new feature is off and that behaves as described in tech spec when feature is on.

The change only extends case search logic and does not affect existing logic. The new logic is gated behind both the feature flag and check that feature is enabled for the module that contains the case type searched. 

There is no effect on existing data. 
 
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Unit tests added that covers:
Pulling expected child cases in variations of Feature Flag on/off and module setting disabled/enabled
Pulling expected parent, child and extension cases

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

[QA Functional Test
](https://dimagi-dev.atlassian.net/browse/QA-4773?focusedCommentId=249104&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249104)
[QA Performance Test
](https://dimagi-dev.atlassian.net/browse/QA-4772?focusedCommentId=249105&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249105)

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

Jtang-1 committed:
refactor: rename

Jtang-1 committed:
refactor: extract getting new related cases

Jtang-1 committed:
refactor: pass only required info into get_child_case_results

Jtang-1 committed:
feat: pulls all related parent, child, and extension cases into case search results

Jtang-1 committed:
feat: adds app managers UI to toggle feature

Jtang-1 committed:
feat: adds pull_parent_child_ext_cases toggle to model

Jtang-1 committed:
feat: grabs pull_parent_child_ext_cases from and to model

Jtang-1 committed:
feat: update pull_parent_child_ext_cases for all modules with the same case type as current module

In case search logic, if any module with a specific case type has the feature on, then searching
that case type will always pull in parent,child, and extension case

Jtang-1 committed:
feat: Uses sync algorithm if FF on and toggled on, otherwise uses detail nodeeset

Jtang-1 committed:
refactor: separate assert in test

Jtang-1 committed:
test: new tests for utilites functions pulling related cases

Jtang-1 committed:
test: patch for new function used in existing test. Function tested was not altered

Jtang-1 commented:
Open to any renaming (function/variables or App Builder Feature name) recommendations

Jtang-1 committed:
fix: also checks module has case search before performing query

Jtang-1 committed:
test: add patches for module_uses_pull_parent_child_ext_cases

snopoke commented on corehq/ex-submodules/casexml/apps/phone/data_providers/case/livequery.py:
```diff
@@ -118,12 +118,16 @@ def get_case_hierarchy(domain, cases):
+    return cases + new_cases
+
+
+def get_parent_and_ext_cases(domain, case_ids):
```
Seems like this change is just to accommodate passing the case IDs instead of the full cases but looking at the callers I think it should be easy enough to pass the full cases. I also think it is necessary because if the app is also using a Data Registry, the cases may not all belong to the same domain and we'd need to segment them. There is already code for doing that, we'd just need to plumb it in:

When creating the search helper we can update the code to return the `DataRegistryHelper` and pass that to the `_RegistryQueryHelper`

```diff
if registry_slug:
-    query_domains = _get_registry_visible_domains(couch_user, domain, case_types, registry_slug)
+    registry_helper = _get_and_validate_registry_helper(couch_user, domain, case_types, registry_slug)
     helper = _RegistryQueryHelper(domain, couch_user, registry_helper)
```

Then in the query helper classes we can add a function:
```python
class _QueryHelper:
    def get_case_hierarchy(self, cases):
        from casexml.apps.phone.data_providers.case.livequery import get_case_hierarchy
        return get_case_hierarchy(self.domain, cases)

class _RegistryQueryHelper:
    def get_case_hierarchy(self, cases):
        return self.registry_helper.get_multi_domain_case_hierarchy(self.couch_user, cases)
```

And then finally use that helper function instead of calling it directly.

Jtang-1 replied:
My main intention of separating it out was actually to be able to grab just the new cases since `get_case_hierarchy` returns both the original and new cases. But `get_parent_and_ext_cases ` (not called `get_parent_host_ext_cases` since commit [da6b314](https://github.com/dimagi/commcare-hq/pull/32587/commits/da6b3145f31536d20533ac734731bc7d07a804f8)) can also take in full cases as a parameter.

Your recommended change makes sense but I'm not sure how the `registry_helper` validation should be handled. For the pseudo `_get_and_validate_registry_helper` function, a `DataRegistryHelper ` is returned if validated, but what should be returned upon a `RegistryNotFound` exception?

snopoke replied:
> My main intention of separating it out was actually to be able to grab just the new cases since `get_case_hierarchy` returns both the original and new cases. But `get_parent_and_ext_cases ` (not called `get_parent_host_ext_cases` since commit [da6b314](https://github.com/dimagi/commcare-hq/pull/32587/commits/da6b3145f31536d20533ac734731bc7d07a804f8)) can also take in full cases as a parameter.

That makes sense. I think the main thing I want to make sure we keep is the domain validation, so if we can move that into the new function as well then 👍🏻 

 
> Your recommended change makes sense but I'm not sure how the `registry_helper` validation should be handled. For the pseudo `_get_and_validate_registry_helper` function, a `DataRegistryHelper ` is returned if validated, but what should be returned upon a `RegistryNotFound` exception?

The error handling can be moved up a level to `get_case_search_results` and if the exception is raised we can just use the normal `_QueryHelper` (the same as if no registry were passed in).

snopoke commented on corehq/apps/app_manager/views/modules.py:
```diff
@@ -1337,13 +1338,22 @@ def _check_xpath(xpath, location):
+def _update_pull_parent_child_ext_cases(app, current_module, new_state: bool):
+    for module in app.get_modules():
+        if module.case_type == current_module.case_type:
+            module.search_config.pull_parent_child_ext_cases = new_state
```
I think I see why but see my comments below where I propose a different solution (putting this setting in the suite XML so that it get's sent with the case search request).

snopoke reviewed (changes requested):
@Jtang-1 thanks for the hard work on this! I've requested some fairly large changes, once you've had a chance to go through the comments let me know if you want to discuss them over voice.

Summary:
1. pass the setting through to case search by putting it in the search request (so that we don't have to try and look it up in the app)
2. we need to be able to handle data registry search requests (across multiple domains)

snopoke commented on corehq/apps/case_search/tests/test_get_related_cases.py:
```diff
@@ -151,17 +155,129 @@ def test_get_related_case_results_expanded_results(self):
+                    result_cases = get_defined_cases(_QueryHelper(self.domain), app, {'teacher'}, SOURCE_CASE_ID)
+                self._assert_case_ids(RESULT_CHILD_ID, result_cases)
+
+    def _assert_related_case_ids(self, cases, paths, expected_case_ids):
```
nice job with the tests!

Jtang-1 commented on corehq/apps/case_search/utils.py:
```diff
@@ -265,22 +270,20 @@ def get_related_cases(helper, app_id, case_types, cases, custom_related_case_pro
         rel for rels in [get_related_case_relationships(app, case_type) for case_type in case_types]
```
I think live query does not handle grandparent cases which is handled by `get_related_case_relationships`

snopoke replied:
It should include grandparents (it will follow the index chain all the way up). Should be simple to verify in `test_get_all_related_cases`.

Jtang-1 replied:
Yes you're right, grandfather cases are fetched with livequery. 
Added to test: [2eab339](https://github.com/dimagi/commcare-hq/pull/32587/commits/2eab339e81c5736dd7c9099f4c4308fbb7b6f340)

Moved that logic to only run if the new feature `get_all_related_cases` is not enabled to run
[4ad0c3a](https://github.com/dimagi/commcare-hq/pull/32587/commits/4ad0c3acd14b32dbdd43be9129e790ef71f52ee0)

Jtang-1 committed:
refactor: rename pull_parent_child_ext_cases to include_related_cases

Jtang-1 committed:
refactor: changes wording displayed on UI for toggle

Jtang-1 committed:
feat: pull include_related_case setting in suite xml to send with case search request

Jtang-1 committed:
refactor: import function directly

Jtang-1 committed:
removes updating applying setting to all modules of same case type

This was done previously because case search didn't know which module the
request is coming from. So setting affects all modules with the same case type.
Now the setting is passed through suite xml and sent with case search request.

Jtang-1 committed:
feat: look up setting through search request instead of through the app

Jtang-1 committed:
refactor: rename function to be more accurate description

Jtang-1 committed:
test: updates for moving setting check to include_related_cases property

Jtang-1 commented on corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_properties.html:
```diff
@@ -364,6 +364,18 @@ <h4 class="panel-title panel-title-nolink">{% trans "Search and Claim Options" %
           </div>
+          <div class="form-group">
+            <label class="control-label {% css_label_class %}">
+              {% trans "Pull All Related Cases" %}
```
[909790e](https://github.com/dimagi/commcare-hq/pull/32587/commits/909790e6612718c7c80dbd378e8bfe79956f8d7b)

Jtang-1 commented on corehq/apps/case_search/utils.py:
```diff
@@ -336,29 +339,64 @@ def get_related_case_results(helper, cases, paths):
+
+def get_parent_child_ext_cases(helper, app, case_types, source_case_ids):
+    get_parent_and_ext_cases = importlib.import_module(
+        "corehq.ex-submodules.casexml.apps.phone.data_providers.case.livequery").get_parent_and_ext_cases
```
Ah, good tip incorporated - [9b20059](https://github.com/dimagi/commcare-hq/pull/32587/commits/9b20059ebca9856810f965bc1f266d96b3a912ec)

Jtang-1 commented on corehq/apps/case_search/utils.py:
```diff
@@ -336,29 +339,64 @@ def get_related_case_results(helper, cases, paths):
+    pull_parent_child_ext_cases = False
+    results = []
+    for module in app.get_modules():
+        if module.case_type in case_types and module_uses_pull_parent_child_ext_cases(module):
```
That makes sense, and I really appreciate the breakdown on how this works. 

Pulling `include_related_case` setting into suite file + reading from it: [ea3e8f0](https://github.com/dimagi/commcare-hq/pull/32587/commits/ea3e8f0f7cd385569578c5706bb3feaa8fd349a3)
[4fe62a9](https://github.com/dimagi/commcare-hq/pull/32587/commits/4fe62a9e25dc9ff7d413e9651a5b1ca96bad3b94)

Removing logic in previous commits that applied change over all modules:[56a4d5a](https://github.com/dimagi/commcare-hq/pull/32587/commits/56a4d5a69aa5996a371cdf588e23be4517ff086d)

Jtang-1 commented on corehq/apps/case_search/utils.py:
```diff
@@ -336,29 +339,64 @@ def get_related_case_results(helper, cases, paths):
+        if module.case_type in case_types and module_uses_pull_parent_child_ext_cases(module):
+            pull_parent_child_ext_cases = True
+            results.extend(get_parent_and_ext_cases(helper.domain, source_case_ids))
+            results.extend(get_child_case_results(helper, source_case_ids))
```
Addressed after implementing the strategy you recommended  - https://github.com/dimagi/commcare-hq/pull/32587/commits/4fe62a9e25dc9ff7d413e9651a5b1ca96bad3b94

Jtang-1 commented on corehq/apps/case_search/utils.py:
```diff
@@ -336,29 +339,64 @@ def get_related_case_results(helper, cases, paths):
     return results
 
 
+def get_defined_cases(helper, app, case_types, source_case_ids):
```
https://github.com/dimagi/commcare-hq/pull/32587/commits/4fe62a9e25dc9ff7d413e9651a5b1ca96bad3b94

snopoke commented on corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:
```diff
@@ -270,6 +273,14 @@ def _remote_request_query_datums(self):
                 )
             )
+        if (module_uses_include_related_cases(self.module)
+                and toggles.USH_CASE_CLAIM_UPDATES.enabled(self.app.domain)):
```
I don't think the toggle is necessary here. There's no issue with backwards compatibility so as long as it's configured in the app then we can add it to the query

Jtang-1 replied:
Removed-  [70b1ed9](https://github.com/dimagi/commcare-hq/pull/32587/commits/70b1ed9cf58d255c14e8062b3874965738fb09ec)

snopoke commented on corehq/apps/case_search/models.py:
```diff
@@ -152,6 +154,7 @@ class CaseSearchRequestConfig:
     case_types = attr.ib(kw_only=True, default=None)
     data_registry = attr.ib(kw_only=True, default=None, converter=_flatten_singleton_list)
     custom_related_case_property = attr.ib(kw_only=True, default=None, converter=_flatten_singleton_list)
+    include_related_cases = attr.ib(kw_only=True, default=None)
```
You should add the converted `converter=_flatten_singleton_list` otherwise this will be a list

Jtang-1 replied:
https://github.com/dimagi/commcare-hq/pull/32587/commits/c4d964d87763c28a1f446949ed18ce6d173b72c8

snopoke commented on corehq/apps/app_manager/views/modules.py:
```diff
@@ -1337,13 +1338,13 @@ def _check_xpath(xpath, location):
     app.save(resp)
     return JsonResponse(resp)
 
-
```
nit: 2 lines between functions is standard

Jtang-1 replied:
fixed with other other renaming - [6fd4217](https://github.com/dimagi/commcare-hq/pull/32587/commits/6fd42170973c08ee8126b8bd5c098e7163c9551a)

snopoke commented on corehq/apps/case_search/utils.py:
```diff
@@ -265,22 +273,20 @@ def get_related_cases(helper, app_id, case_types, cases, custom_related_case_pro
 
-    if child_case_types:
-        results.extend(get_child_case_results(helper, top_level_cases, child_case_types))
+    defined_cases = get_defined_cases(helper, app, case_types, top_level_case_ids, include_related_cases)
```
what does `defined` mean in this context?

Jtang-1 replied:
There were overlapping meanings of "get related cases" so I renamed them to essentially be:
High Level: `get_related_cases_result` (previously `get_defined_cases`

new feature: `_get_all_related_cases`
else: 1.`_get_search_detail_path_defined_cases` (previously named `get_related_case_results`) and 2. `_get_child_cases_referenced_in_app`

[851645a](https://github.com/dimagi/commcare-hq/pull/32587/commits/851645acd35d3d0ab5762ef780cfceb61fd176d6)
[6fd4217](https://github.com/dimagi/commcare-hq/pull/32587/commits/6fd42170973c08ee8126b8bd5c098e7163c9551a)

snopoke reviewed (commented):
Changes are looking good so far.

Jtang-1 committed:
remove redundant FF toggle check. Check that setting is configured in app is enough

Jtang-1 committed:
property should not be a list

Jtang-1 committed:
feat: makes changes to work for Data Registry helpers

Jtang-1 committed:
pass cases instead of just case ids to accomodate helper

Jtang-1 committed:
test: pass cases into get_related_case instead of case ids

Jtang-1 committed:
test: tests get_related_case_results path where include_related_cases is True

Jtang-1 committed:
test: adds check that grandfather cases are fetched

Jtang-1 committed:
feat: cases fetches from search_detail_path_defined_cases already fetched from get_all_related_cases

Jtang-1 committed:
test: repatch for get_search_detail_path_defined_cases moved into get_all_related_cases

Jtang-1 committed:
refactor: rename

Jtang-1 committed:
refactor: reorder functions and make private

Jtang-1 committed:
refactor: rename for clarity and delete unused imports

Jtang-1 committed:
correct test failure for case search registry

Jtang-1 committed:
correct variable names for test

Jtang-1 committed:
lint

snopoke commented on corehq/apps/case_search/utils.py:
```diff
@@ -261,34 +275,71 @@ def get_related_cases(helper, app_id, case_types, cases, custom_related_case_pro
 
-    if child_case_types:
-        results.extend(get_child_case_results(helper, top_level_cases, child_case_types))
+    defined_cases = get_related_cases_result(helper, app, case_types, top_level_cases, include_related_cases)
```
nit: rename `defined_cases` to `related_cases`

Jtang-1 replied:
renamed [ccc06cf](https://github.com/dimagi/commcare-hq/pull/32587/commits/ccc06cf9364a69edfb8ffa924042c56afb4baa3e)

snopoke commented on corehq/apps/case_search/utils.py:
```diff
@@ -261,34 +275,71 @@ def get_related_cases(helper, app_id, case_types, cases, custom_related_case_pro
+    result = []
+    if child_case_types:
+        result.extend(get_child_case_results(helper, source_case_ids, child_case_types))
+    return result
```
nice refactoring

snopoke commented on corehq/apps/case_search/utils.py:
```diff
@@ -343,22 +394,23 @@ def get_child_case_types(app, case_type):
-               .get_child_cases(parent_case_ids, "parent")
-               .run().hits)
+def get_child_case_results(helper, parent_case_ids, child_case_types=None):
+    filter = (helper.get_base_queryset()
+            .get_child_cases(parent_case_ids, "parent"))
```
nit: the formatting is a bit off here. I think this can go on one line now too.
```suggestion
    filter = helper.get_base_queryset().get_child_cases(parent_case_ids, "parent")
```

Jtang-1 replied:
Yes, corrected [ccc06cf](https://github.com/dimagi/commcare-hq/pull/32587/commits/ccc06cf9364a69edfb8ffa924042c56afb4baa3e)

snopoke reviewed (commented):
This is looking great. Mostly minor comments.

Jtang-1 commented on corehq/apps/case_search/utils.py:
```diff
@@ -95,15 +91,19 @@ def get_case_search_results(domain, case_types, criteria,
+        else:
+            query_domains = registry_helper.visible_domains
+            helper = _RegistryQueryHelper(domain, couch_user, registry_helper)
+    return helper, query_domains
```
by expose, do you mean splitting this into two functions, a `get_helper` and `get_visible_domains`?

snopoke replied:
I mean making a property on the helper classes `helper.domains`

Jtang-1 replied:
Got it - see [aeba559](https://github.com/dimagi/commcare-hq/pull/32587/commits/aeba5590cb7078037ad1228e6f0786591f5fca50)

Jtang-1 committed:
datum in this case acts only as boolean flag. If this datum exists, then it is always true

Jtang-1 committed:
test: removes unused instance of app

Jtang-1 committed:
minor format and variable name change

Jtang-1 commented on corehq/apps/case_search/tests/test_get_related_cases.py:
```diff
@@ -149,19 +164,124 @@ def test_get_related_case_results_expanded_results(self):
+    def test__get_all_related_cases(self):
+        app = Application.new_app(self.domain, "Case Search App")
+        module = app.add_module(Module.new_module("Search Module", "en"))
+        module.case_type = "teacher"
```
no it's not, removed [a5f6ec2](https://github.com/dimagi/commcare-hq/pull/32587/commits/a5f6ec269fd49843558086a727dc10239ee385ff)

Jtang-1 commented on corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:
```diff
@@ -270,6 +273,13 @@ def _remote_request_query_datums(self):
+            datums.append(
+                QueryData(
+                    key=CASE_SEARCH_INCLUDE_RELATED_CASES_KEY,
+                    ref=f"'{self.module.search_config.include_related_cases}'",
```
That makes sense - [aeba559](https://github.com/dimagi/commcare-hq/pull/32587/commits/aeba5590cb7078037ad1228e6f0786591f5fca50)

Jtang-1 committed:
refactor: expose query_domains as class property instead of returned in tuple

Jtang-1 committed:
refactor: rename for clarity. feature controls getitng all related cases. Related cases are fetched through other logic even if this setting is off

Jtang-1 committed:
squash with previous

Jtang-1 commented on corehq/apps/case_search/tests/test_get_related_cases.py:
```diff
@@ -81,47 +84,59 @@ def test_get_related_case_results(self):
         self.assertEqual(max(case_ids.values()), 1, case_ids)  # no duplicates
 
-    def test_get_related_case_results_expanded_results(self):
+        with patch("corehq.apps.case_search.utils.get_search_detail_relationship_paths",
```
Tried to avoid  comments if possible- I separated out calls of `get_related_cases` into a helper function that only takes in the bool `include_all_related_cases` param to make it more obvious. Otherwise, extracted it into a variable.

Can add comments if you think it's still not clear enough
[f1210d9](https://github.com/dimagi/commcare-hq/pull/32587/commits/f1210d96f59b4e13cb74dc4aee78daf78c4f6df0)

Jtang-1 committed:
refactor: rename variables for clarity and reorganize to make include_all_related_cases parameter more evident

Jtang-1 commented on corehq/apps/case_search/tests/test_get_related_cases.py:
```diff
@@ -149,19 +164,124 @@ def test_get_related_case_results_expanded_results(self):
+        result_cases = get_child_case_results(_QueryHelper(self.domain), SOURCE_CASE_ID, CHILD_CASE_TYPE_FILTER)
+        self._assert_case_ids(WITH_FILTER_RESULT_CHILD_CASE_ID, result_cases)
+
+    def test__get_all_related_cases(self):
```
Added high level comment. Also explicitly defined the expected results in a dict to better interpret the cases. 

[ef39518](https://github.com/dimagi/commcare-hq/pull/32587/commits/ef3951824c8babef253fb6c798678c96cd68cffd)

Jtang-1 committed:
lint

snopoke commented on corehq/apps/case_search/utils.py:
```diff
@@ -343,22 +393,22 @@ def get_child_case_types(app, case_type):
-               .get_child_cases(parent_case_ids, "parent")
-               .run().hits)
+def get_child_case_results(helper, parent_case_ids, child_case_types=None):
+    filter = (helper.get_base_queryset().get_child_cases(parent_case_ids, "parent"))
```
nit: redundant parenthesis 
```suggestion
    filter = helper.get_base_queryset().get_child_cases(parent_case_ids, "parent")
```

Jtang-1 replied:
[26e3ffa](https://github.com/dimagi/commcare-hq/pull/32587/commits/26e3ffaa29ec122db06e7e0f56f7d75381d7d213)

snopoke commented on corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_properties.html:
```diff
@@ -364,6 +364,18 @@ <h4 class="panel-title panel-title-nolink">{% trans "Search and Claim Options" %
+                </label>
+              </div>
+            </div>
+          </div>
```
Is this being bundled under `USH_CASE_CLAIM_UPDATES`?

Jtang-1 replied:
Yes it is, this change will be under USH_CASE_CLAIM_UPDATES FF

snopoke reviewed (approved):
Looking great!

Jtang-1 committed:
bug: tags "get_all_related_live_cases" as "related_cases" to filter from case list display.
refactor: moves tagging cases from elastic search wrapper to case search utils

Jtang-1 committed:
naming: rename to capture tagging side effect

Jtang-1 committed:
Merge branch 'master' into jt/ush_2575_case_search_pull_related_cases

Jtang-1 committed:
lint and add test

Jtang-1 commented:
@snopoke  the main change since you last reviewed is [ec21fd7](https://github.com/dimagi/commcare-hq/pull/32587/commits/ec21fd770eada9cee84eaec815c5618644bc33cd). This corrects a bug QA caught where the pulled related cases were displayed on case list if the case types are defined in "Additional Case List and Case Search Type".  Can I get your review on this change?

Jtang-1 committed:
Merge branch 'master' into jt/ush_2575_case_search_pull_related_cases

Jtang-1 commented:
Merged, thank you again for the thorough review and feedback on this PR Simon!
